### PR TITLE
Implement auto-scaling scripts and refactor 

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
Partitions can optionally define three new scripts:
* `grow_script` => Ran when a node's reason is set to `Resources`,
* `shrink_script` => Ran if a node is idling after a reallocation message is received, and
* `status_script` => Ran periodically

All the scripts are ran debounced to prevent them running to frequently. This is controlled by a minimum/maximum config value. The message will be sent after:
* the minimum period has elapsed without an additional request to send the message (of the same type),
* the maximum time has elapsed, or
* somewhere in between.

NOTE:
* The `status_script` is sent at a regular frequency which is controlled by an independent config. This makes debouning it a moot point. It has still be implemented in case other triggers are added in the future
* Debouncing is controlled at the partition level, not application wide. This means multiple message maybe sent from different partitions.

----

This work required an upgrade to the partitions syntax which highlighted the need to improve how partitions are validated. Previously it was done on an ad-hoc bases as the partitions are parsed. This caused various issues including:
* Lack of uniform error handling,
* Poor handling of incorrect syntax,
* Non atomic acceptance/rejection of the config,
* Week guarantees on types, and
* Proliferation of partition code in the core configuration.

As such the `FlightScheduler::Configuration` has been refactored to remove most of the `partition` related code. Most of this has been relocated into a `Partition::Builder` class which uses [JSON Schema](https://json-schema.org/understanding-json-schema/about.html) to validate the user provided config. This has a few advantages:
* It validates exactly what the user provided,
* It can not fail to generate an intermediary internal object before validation,
* It gives rich error information including the exact location it occur, and
* It can easily handle multiple error conditions within the config.

Its main downside is its exactness is a double edge sword. It produces an error message which is more machine readable then human readable. This can be addressed in time with an error message generator, but has not been implemented.

It also can not validate certain higher order conditions such as uniqueness and file existence. These validations are done using the standard `ActiveMode::Validations` module. These validations are done after the config has been loaded to allow access to `FlightScheduler.app,config` values. Without which the file paths can not be generated.

---

Finally a lot of the `node`-`partition` mapping code has been moved onto the `NodeRegistry`. This provides a single location that a node needs to update after being profiled. There was to allow for the persistence of dynamic nodes, however this wasn't required in the end.

### CODE REVIEW

File: app/models/allocation.rb [hunk 1]
```diff
@@ -51,7 +53,11 @@ def initialize(job:, nodes:)
 
   def nodes
     @nodes ||= @node_names.map do |node_name|
-      FlightScheduler.app.nodes.fetch_or_add(node_name)
+      FlightScheduler.app.nodes[node_name].tap do |node|
+        raise MissingNodeError, <<~ERROR.chomp if node.nil?
+          Tried to allocate missing node: '#{node_name}'
+        ERROR
+      end
```
Comment (benarmston) [inline app/models/allocation.rb]: Previously, this was potentially adding nodes to the registry to support the following:

1. Controller is running and a job is allocated.
2. The allocation registry is persisted.
3. The controller is restarted whilst there are still allocations in the registry.
4. The allocation registry state file is read and the allocations created.
5. The nodes connect to the controller.

I found that there was a need for the nodes to be created in this case.  Otherwise certain things would break, but right now I can't recall exactly what broke.

Currently, a non-empty `allocation_state` prevents the controller from restarting.

File: app/models/partition.rb [hunk 2]
```diff
@@ -25,42 +25,353 @@
...
+  def dynamic?
+    grow_script_path ? true : false
+  end
+
+  # NOTE: This method considers the partition shrinkable if any nodes are IDLE
+  # Some additional handling maybe required for node's which are DOWN but not terminated
+  # OR if IDLE static nodes count
+  def shrinkable?
+    dynamic? && nodes.any? { |n| n.state == 'IDLE' }
+  end
```
Comment (benarmston) [inline app/models/partition.rb]: Shouldn't this test the existence of the shrink script, not the grow script.

File: lib/flight_scheduler/node_matcher.rb [hunk 3]
```diff
@@ -35,27 +35,37 @@ class NodeMatcher
...
       @key = key.to_s
       @specs = specs.transform_keys(&:to_s)
     end
 
+    # DEPRECATED: These methods are used extensively in the specs but should not
+    # be used in the code base. They have been replaced by the top level partition
+    # config validation.
+    #
+    # Consider refactoring
+    attr_reader :errors
```
Comment (benarmston) [inline lib/flight_scheduler/node_matcher.rb:66]: Can you explain what this means and why we're leaving it in if its "deprecated"?  Why not just remove it?

File: lib/flight_scheduler/node_registry.rb [hunk 4]
```diff
@@ -25,38 +25,73 @@
...
       end
-    end
-  end
 
-  def [](node_name)
-    with_lock { @nodes[node_name] }
-  end
+      @partitions_cache.each do |partition, nodes|
+        match     = partition.node_match?(node)
+        existing  = nodes.include?(node)
```
Comment (benarmston) [inline lib/flight_scheduler/node_registry.rb]: What stops `nodes` being `nil` here?

File: lib/flight_scheduler/node_registry.rb [hunk 5]
```diff
@@ -25,38 +25,73 @@
...
-        yield
+      # Return the node
+      node
+    end
+  end
+
+  def for_partition(partition)
+    # The cache update is atomic and solely dependent of the state of 'nodes'
+    # As such it can be done in a read lock
+    @lock.with_read_lock do
```
Comment (benarmston) [inline lib/flight_scheduler/node_registry.rb]: I'm not sure its as atomic as you think it is.  We're potentially creating a new key in the hash which IIRC could result in the hash updating internal references in a non-threadsafe manner.  The use of `with_read_lock` and `with_write_lock` ensures that `update_node` and `for_partition` do not interleave in an unsafe manner, but I can't see anything that would prevent multiple calls to `for_partition` from interleaving in an unsafe manner.

File: spec/models/partition_spec.rb [hunk 6]
```diff
@@ -28,10 +28,97 @@
...
 RSpec.describe Partition, type: :model do
-  let(:job) {
-    Job.new(
-      id: 1,
-      min_nodes: '2',
-    )
-  }
+end
+
+RSpec.describe Partition::Builder do
```
Comment (benarmston) [inline spec/models/partition_spec.rb]: This would be a lot better being in its own file.  Perhaps `partition/builder_spec.rb` or `partition_builder_spec.rb`.  Long spec files become very hard to read or navigate in my opinion.

File: etc/flight-scheduler-controller.yaml [hunk 7]
```diff
@@ -90,13 +90,29 @@
...
+# status_update_period: 3600
+
+# The minimum and maximum debouncing periods in seconds
+# Environment variables FLIGHT_SCHEDULER_MIN_DEBOUNCING and FLIGHT_SCHEDULER_MAX_DEBOUNCING
+# takes precedence over this setting.
+# Min Default: 30
+# Max Default: 600
+# min_debouncing: 5
+# max_debouncing: 15
+
```
Comment (benarmston) [inline etc/flight-scheduler-controller.yaml]: We should have an entry for `libexec` here too.

File: app/models/partition.rb [hunk 8]
```diff
@@ -25,42 +25,353 @@
...
+
+    def run(path, type:)
+      if path.nil?
+        Async.logger.debug "Skipping #{type} script for partition #{partition.name} as it does not exist"
+        return
+      end
+
+      Async.logger.info "Running (#{type}): #{path}"
+      stdin_str = JSON.pretty_generate(stdin)
+      Async.logger.debug "STDIN: #{stdin_str}"
```
Comment (benarmston) [inline app/models/partition.rb]: This should be changed to `Async.logger.debug("STDIN:") { stdin_str }`.  Passing the stdin string as a block have the async logger perform some pretty pretty printing in the log file making it much easier to follow when the log entry starts and ends.

File: app/models/partition.rb [hunk 9]
```diff
@@ -25,42 +25,353 @@
...
+        #{out}
+
+        STDERR:
+        #{err}
+      MSG
+      if status.success?
+        Async.logger.info msg
+      else
+        Async.logger.warn msg
+      end
```
Comment (benarmston) [inline app/models/partition.rb]: Similar here.  Can we get multi-line strings passed to the logger as a block.  This should be a general rule when using the logger as it makes scanning it much easier.

File: app/models/partition.rb [hunk 10]
```diff
@@ -25,42 +25,353 @@
...
+        "required" => ["grow_script", "shrink_script", "status_script"],
+        "properties" => {
+          "grow_script" => { "type" => "string" },
+          "shrink_script" => { "type" => "string" },
+          "status_script" => { "type" => "string" }
+        }
+      }
+    },
+    "if" => { "properties" => { "dynamic" => { "type" => "null" } } },
+    "else" => { "required" => ["node_matchers"] }
```
Comment (benarmston) [inline app/models/partition.rb]: Can you explain what these lines do.

**Review by benarmston — COMMENTED** [review]
## The implementation is too focussed on dynamic partitions

My general thoughts on this PR is that it is too focussed on "dynamic" partitions, rather than sending notifications.  Instead of thinking about a mechanism to grow or shrink the partition we should be thinking of running a script in response to the scheduler experiencing some event.

The events that the scheduler experiences are:

1. There are jobs I can't schedule because they are waiting on resources.
2. There are idle resources for which no jobs are waiting.


There is also another event which you've captured well.

3. A period of time has passed, this is the status of your partitions.

The areas where thinking about "dynamic" partitions have bled into the code in a way that doesn't seem quite right are:

1. The partition configuration key `dynamic`.  This would be better named `event_scripts` or something similar.

2. The name of the scripts.  Instead of `grow` and `shrink` these would be better named after the event.  Perhaps, `insufficient_resources` and `excess_resources`.

3. The partition methods `dynamic?` and `shrinkable?`.  Partitions aren't growable or shrinkable.  They have nodes which may or may not be turned off or turned on in response to some event script running.

The above might seem pedantic, and to some extent it is.  However, Steve seems to have a path of how to direct this scheduler to something that our customers are going to be familiar with / want to use whilst still supporting cloud based clusters.  It's important that we don't deviate too far from that or we may well end up in the position that we have to deal with our deviation being incompatible with future requests.


## The scheduler is not sufficiently involved in deciding when to run the scripts

I also think that the scheduler is not sufficiently involved in the decision of when to run the scripts.  Take for instance the `shrink`/`excess_resources` script.  Currently, this is scheduled to run whenever a job is deallocated on any of its nodes.  If a job is running on multiple nodes, the script will scheduled to run multiple times.  There is a check that the partition is `shrinkable?`, but the partition cannot determine that correctly.  It doesn't know what the status of the queue is.

Consider that there is an `ARRAY_JOB` with many tasks each requiring two nodes.  All of the partition's resources are turned on/connected and they are all allocated to tasks from the `ARRAY_JOB`.  When one of the tasks completes, it will receive two `NODE_DEALLOCATED` messages; one from each node that was assigned to the task.  For the first node to send the `NODE_DEALLOCATED` message, the following sequence happens.


1. Attempt to allocate more jobs. Fail to do so as all jobs require two nodes and only a single node is available.
2. Schedule the `grow`/`insufficient resources` script to run, as we failed to allocate a job.
3. Determine that the partition can be shrunk, as it has idle nodes.
4. Schedule the `shrink`/`excess resources` script to run.

This will result in the `shrink` script being ran when it shouldn't.  Depending on what jobs are pending, it can also result in the `shrink` script being ran immediately prior to or after the `grow` script being ran.

Running the `shrink` script won't necessarily result in the admin being messaged as the `shrink` script is provided enough information to determine that the event "There are idle resources for which no jobs are waiting" has not actually happened.  However, there are issues with having the script make that determination.  Namely, they won't all make it or won't all make it consistently.  If the scheduler or some other class makes that determination we will have consistent behaviour.


## Other thoughts:

The partition file is looking very large now.  It would be worth extracting the `Builder` and `ScriptRunner` to `app/models/partition/builder.rb` and `app/models/partition/script_runner.rb`.

I've previously said that I don't like extracting single-use concerns, because to me a concern is supposed to be generic enough to be mixed into many classes.  If it isn't, I think that is most explicitly communicated by co-locating it with the single class that it works with.

However, for `Partition::Builder` and `Partition::ScriptRunner` it is clear from the name of the classes that they intended to work with the `Partition` class and not any class.  Moving them out to their own file when they are as large as they are is generally a good thing in my opinion.

One for future consideration; an interesting consistency change might be to create a "static nodes" matcher.  Which matches against a static list of node names.  That might clear up some of the special handling of static nodes.  We could expose such a matcher to the partition YAML syntax or not. Not one for now but it might be worth considering that. 


---

The following are needed to get this PR to a state where it can be merged:

1. Fix the issue that prevents the controller from restarting if there is a non-empty `allocation_state` file.

2. Rename the `dynamic` partition configuration section and scripts.

3. Fix the issue where the `shrink` script can be ran incorrectly for multi-nodes jobs.

4. Minor easy changes such as log formatting.

The other issues can be left for later.  This includes moving decision making out of the script and into the scheduler, and extracting `Builder` and `ScriptRunner` (unless that is easy).

File: app/models/partition.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply app/models/partition.rb]: Whether `dynamic?` tests the existence of the "grow"/"shrink" script is a moot point. There is a validation to ensure all three scripts ("grow"/"shrink"/"status") are provided or they are not.

So ATM testing one is testing all of them

File: lib/flight_scheduler/node_matcher.rb (hunk 3 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/node_matcher.rb:66]: Currently it used in the `node_matcher_spec` file in a few places. However the node matchers are no longer validated individually as this is done on config load.

To remove the methods would require refactoring the specs, which I decided could be left for another time.

File: lib/flight_scheduler/node_registry.rb (hunk 4 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/node_registry.rb]: It is now guaranteed not to be nil
https://github.com/openflighthpc/flight-scheduler-controller/commit/fabb7ddce036bdf9293dcf734f72b130b91623fb

File: etc/flight-scheduler-controller.yaml (hunk 7 above)
Comment (WilliamMcCumstie) [reply etc/flight-scheduler-controller.yaml]: https://github.com/openflighthpc/flight-scheduler-controller/commit/0f60fab373f77818458b817d1f9a74c5b5ba9637

File: app/models/partition.rb (hunk 8 above)
Comment (WilliamMcCumstie) [reply app/models/partition.rb]: https://github.com/openflighthpc/flight-scheduler-controller/commit/64a0f586dde4981a94e5c1483dc59314891bcc46

File: app/models/partition.rb (hunk 9 above)
Comment (WilliamMcCumstie) [reply app/models/partition.rb]: https://github.com/openflighthpc/flight-scheduler-controller/commit/e1da3bb7f65c92b8d0bfed6f8a49056ba3153be2

File: app/models/partition.rb (hunk 10 above)
Comment (WilliamMcCumstie) [reply app/models/partition.rb]: As discussed IRL, they forced the `node_matcher` flag to be specified with the `dynamic` option. However this is not required and has been removed:
https://github.com/openflighthpc/flight-scheduler-controller/commit/9d5e67ed1e1f857ed85db3b04bcf6501b7652f8a

File: app/models/allocation.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply app/models/allocation.rb]: In general, creating an allocation with a missing node is an error condition. However an optional `add` flag has been added to handle the above edge case due to the `persistence`: https://github.com/openflighthpc/flight-scheduler-controller/commit/b8e01726e5f35698db1a01ee0a8947a6c538c4db

I can't think of any other situations this will occur, as such they can be dealt with if and when a problem arises.

File: lib/flight_scheduler/node_registry.rb (hunk 5 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/node_registry.rb]: Ah fair point, a write lock has been added:
https://github.com/openflighthpc/flight-scheduler-controller/commit/f4f525cfc5aae6034928244b91563a1a52a5e383

File: app/models/partition.rb (hunk 2 above)
Comment (benarmston) [reply app/models/partition.rb]: Fair enough.  I'd looked for such validation and hadn't found it.  And now that I know it *is* there I can find it easily :-)

File: app/models/allocation.rb (hunk 1 above)
Comment (benarmston) [reply app/models/allocation.rb]: You've used `map` where I think you should have used `tap`.  It's still broken.

File: lib/flight_scheduler/node_matcher.rb (hunk 3 above)
Comment (benarmston) [reply lib/flight_scheduler/node_matcher.rb:66]: Fair enough.

File: lib/flight_scheduler/node_registry.rb (hunk 4 above)
Comment (benarmston) [reply lib/flight_scheduler/node_registry.rb]: I was thinking more that the value of the `@partitions_cache` hash could be `nil`.  But this code block assumes that it is always an `Array`.  I don't think that is a valid assumption.

Comment (benarmston) [reply lib/flight_scheduler/node_registry.rb]: Nope.  I'm mistaken here.  If a partition has been added to the cache it has been added with an `Array` of nodes.

File: app/models/partition.rb (hunk 8 above)
Comment (benarmston) [reply app/models/partition.rb]: Thank you.

File: app/models/allocation.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply app/models/allocation.rb]: Good spot, turns out it is neither:
https://github.com/openflighthpc/flight-scheduler-controller/commit/e5d61701254720251a1138aaba71e0abb2fe32fc

File: lib/flight_scheduler/node_registry.rb (hunk 4 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/node_registry.rb]: :+1:

**Comment by WilliamMcCumstie** [comment]
Rebased to resolve a few spec issues

File: spec/models/partition_spec.rb (hunk 6 above)
Comment (WilliamMcCumstie) [reply spec/models/partition_spec.rb]: https://github.com/openflighthpc/flight-scheduler-controller/commit/2c73822981a90920c8c79d759068ca4000e29772

**Comment by WilliamMcCumstie** [comment]
Rebased on master to resolve conflict

**Comment by benarmston** [comment]
Does that rebase mean this is good for review again?  Or should I hold off for a bit?

**Comment by WilliamMcCumstie** [comment]
Yes it is ready for review

File: app/models/partition.rb [hunk 11]
```diff
@@ -25,42 +25,134 @@
...
+
+  def insufficient_script_path
+    return nil unless @insufficient_script
+    @insufficient_script_path ||= File.expand_path(@insufficient_script, FlightScheduler.app.config.libexec_dir)
+  end
+
+  def status_script_path
+    return nil unless @status_script
+    @status_script_path ||= File.expand_path(@status_script, FlightScheduler.app.config.libexec_dir)
+  end
```
Comment (benarmston) [inline app/models/partition.rb:120]: These methods don't seem quite right here.  Apart from the validations, `ScriptRunner` is the only class that uses the methods.  Moving the validations to the `ScriptRunner` might be a better solution.

Not one to worry about now though.

File: app/websocket_app.rb [hunk 12]
```diff
@@ -259,7 +259,7 @@ class WebsocketApp
   end
 
   def update_node(node_name, message)
-    node = FlightScheduler.app.nodes.fetch_or_add(node_name)
+    node = FlightScheduler.app.nodes.update_node(node_name)
```
Comment (benarmston) [inline app/websocket_app.rb]: This code had me confused for a while.  It looks like it updates the node before it has updated the node's attributes and then doesn't update the node after the attributes have been updated.  I see now that `Node#attributes=` updates the partitions that the node belongs to.  In which case this call to `NodeRegistry#update_node` is being used to either fetch or add the node.  The naming of the methods isn't quite right here.

Again, this can wait for another time.

File: config/initializers/persistence.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on FlightSchedulerController, please visit:
+# https://github.com/openflighthpc/flight-scheduler-controller
+#==============================================================================
+
+FlightScheduler.app.job_registry.load
+FlightScheduler.app.allocations.load
+FlightScheduler.app.init_periodic_processors
```
Comment (benarmston) [inline config/initializers/persistence.rb]: We've lost the guard against not running this during tests.  When controller specs are ran, the registries are loaded with whatever state has been persisted potentially making them non-deterministic.  The periodic processors could also result in unexpected behaviour in tests.  Please add the guard back.

File: lib/flight_scheduler/job_registry.rb [hunk 14]
```diff
@@ -130,6 +130,16 @@ def load
...
+    jobs.each do |job|
+      # NOTE: There is an edge case where the "insufficient script" would have otherwise
+      # ran before the reboot, but not after. This can occur if the default reason
+      # in the scheduler is changed to anything other than 'Resources'
+      #
+      # This does not happen in the FiFoScheduler and BackfillingScheduler, so can
+      # be ignored for the time being.
+      # Consider revisiting when this no longer holds
+      job.partition.script_runner.insufficient if job.reason_pending == 'Resources'
+    end
```
Comment (benarmston) [inline lib/flight_scheduler/job_registry.rb:142]: I don't understand why we need this.  What would be wrong with simply waiting until we run the next allocation loop?

File: lib/flight_scheduler/schedulers/fifo_scheduler.rb [hunk 15]
```diff
@@ -69,6 +69,13 @@ def run_allocation_loop(candidates)
...
-    new_allocations
+    excess = partition.nodes.any? { |n| n.state == 'IDLE'}
+    [new_allocations, excess]
+  end
+
+  def excess_resources?(partition)
+    FlightScheduler.app.job_registry.jobs.any? do |job|
+      job.partition == partition && job.reason_pending == 'Resources'
+    end
   end
```
Comment (benarmston) [inline lib/flight_scheduler/schedulers/fifo_scheduler.rb]: What's this method for?  It doesn't seem to be called from anywhere.

File: lib/flight_scheduler/schedulers/backfilling_scheduler.rb [hunk 16]
```diff
@@ -75,6 +75,9 @@ def run_allocation_loop(candidates)
       break
     end
 
+    # Works out if there are more idling nodes then the shortage
+    excess = (partition.nodes.count { |n| n.state == 'IDLE' } > backfill.shortage)
```
Comment (benarmston) [inline lib/flight_scheduler/schedulers/backfilling_scheduler.rb]: This isn't quite right.  Consider the case where 2 nodes have connected to the controller.  There are 2 jobs requiring 2 nodes and a third job requiring 3 nodes.  When the second job is completed, the shortage is calculated as `1`, the number of idle nodes is calculated as `2`.  This is determined to be an excess of nodes and so an `excess` notification is sent.  Also, shortly later, the third job is set as pending on `Resources`, so an `insufficient` notification is sent.

As the backfilling is currently changing significantly, getting this correct right now is probably not going to be correct when my changes  land.  We should change this to, "if there are zero jobs pending on resources and there are idle nodes we have an excess".  When my backfilling changes land, it can be looked at again then.

File: lib/flight_scheduler/schedulers/fifo_scheduler.rb [hunk 17]
```diff
@@ -69,6 +69,13 @@ def run_allocation_loop(candidates)
       .select { |job| job.reason_pending == 'WaitingForScheduling' }
       .each { |job| job.reason_pending = 'Priority' }
 
-    new_allocations
+    excess = partition.nodes.any? { |n| n.state == 'IDLE'}
```
Comment (benarmston) [inline lib/flight_scheduler/schedulers/fifo_scheduler.rb]: This is not quite right either.  Consider the case where there is one node connected to the controller.  The queue is empty.  A job requiring 2 nodes is added.  An insufficient message is sent, immediately followed by an excess message.

This check should be changed to the same as the backfilling check: "if there are zero jobs pending on resources and there are idle nodes we have an excess otherwise we don't"

**Review by benarmston — COMMENTED** [review]
This looks like its moving in the right direction.  I've found a couple of issues where running the allocation loop can result in both `excess` and `insufficient` messages being sent.  There are also a few more comments/questions.

File: config/initializers/persistence.rb (hunk 13 above)
Comment (WilliamMcCumstie) [reply config/initializers/persistence.rb]: No need for the guard as `config/initializers/persistence.rb` as required in `config.ru`. Requiring `config.ru` breaks rspec as the rack DSL hasn't been loaded. So there is a relatively low chance of issues in the future.

Regardless, the guard assumes rspec sets the RACK_ENV correctly before being required which could also be broken

File: lib/flight_scheduler/job_registry.rb (hunk 14 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/job_registry.rb:142]: Because an `insufficient` request could be in the denouncing loop whilst the service is being reloaded, which could cause it to be missed. The allocation loop does not run at regular intervals, so we can not assume it will pick up the missed message.

File: lib/flight_scheduler/schedulers/fifo_scheduler.rb (hunk 15 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/schedulers/fifo_scheduler.rb]: It's junk from an implementation I scrapped. Removed: https://github.com/openflighthpc/flight-scheduler-controller/pull/42/commits/fa669934fc0b336f566044a9fa1fee62dedc5caa

File: app/websocket_app.rb (hunk 12 above)
Comment (WilliamMcCumstie) [reply app/websocket_app.rb]: Figured you where going to say that :sweat_smile: 

During the refactor I found there was a corollary between  `update_nodes` and `fetch_or_add`. The `update_nodes` method won because it provides the super set of required features.

But I do agree, the naming could do work

File: app/models/partition.rb (hunk 11 above)
Comment (WilliamMcCumstie) [reply app/models/partition.rb:120]: I'm about to add another `Partition::Type` class which also has validation concerns. They can probably be addressed at the same time

File: lib/flight_scheduler/job_registry.rb (hunk 14 above)
Comment (benarmston) [reply lib/flight_scheduler/job_registry.rb:142]: So that's the problem that we're working around here.  We need the allocation loop to run at regular intervals.  We'll get to that in another PR.

File: lib/flight_scheduler/schedulers/fifo_scheduler.rb (hunk 17 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/schedulers/fifo_scheduler.rb]: https://github.com/openflighthpc/flight-scheduler-controller/commit/7e2b132082203df553a1e162e4f0d2e864539156

File: lib/flight_scheduler/schedulers/backfilling_scheduler.rb (hunk 16 above)
Comment (WilliamMcCumstie) [reply lib/flight_scheduler/schedulers/backfilling_scheduler.rb]: https://github.com/openflighthpc/flight-scheduler-controller/commit/7e2b132082203df553a1e162e4f0d2e864539156

**Comment by WilliamMcCumstie** [comment]
I've made those changes you requested. It is still possible for `insufficient` and `excess` to be sent in short succession due to how the debouncing works. However this can probably be improved upon

File: app/websocket_app.rb (hunk 12 above)
Comment (WilliamMcCumstie) [reply app/websocket_app.rb]: I've renamed `update_node` to be `register_node` as it best reflects the purposes of the method:
1. Ensure the node by the given `name` exists within the registry, and
2. Ensure the internal partition cache is up to date for the node.

There is no longer an `add` flag on this method; as in practice the nodes _should_ always exist. The edge cases where they don't only occur in the specs. As such this is handled in `Node#attribute=` method pending refactoring.

https://github.com/openflighthpc/flight-scheduler-controller/pull/42/commits/81e926f8afec3d07d4a918fa3437886a8c91af19

**Comment by benarmston** [comment]
Rebased on master to fix conflicts.

File: config/initializers/persistence.rb (hunk 13 above)
Comment (benarmston) [reply config/initializers/persistence.rb]: I've added the guard back.  It doesn't change the behaviour but it does document that these actions should not happen under the test environment.  Explicitly documenting this next to the code is preferable.

**Comment by benarmston** [comment]
I've made a few minor changes and this now LGTM.
